app/Platform/Commands/UpdatePlayerGameMetrics.php

```diff
@@ -28,40 +31,63 @@ public function handle(): void
+            $gameIds = collect(explode(',', $this->argument('gameIds') ?? ''))
+                ->filter()
+                ->map(fn ($id) => (int) $id);
 
-        $progressBar = $this->output->createProgressBar($playerGames->count());
-        $progressBar->start();
+            $user = is_numeric($userId)
+                ? User::findOrFail($userId)
+                : User::where('User', $userId)->firstOrFail();
 
-        // if running in sync mode, just call updatePlayerMetrics once manually after updating
-        // all of the player_games instead of letting it cascade for every player_game updated.
-        $isSync = (config('queue.default') === 'sync');
+            $query = $user->playerGames()
+                ->with(['user', 'game']);
 
-        foreach ($playerGames as $playerGame) {
-            $this->updatePlayerGameMetrics->execute($playerGame, silent: $isSync);
-            $progressBar->advance();
-        }
+            if ($gameIds->isNotEmpty()) {
+                $query->whereIn(
+                    'game_id',
+                    Game::whereIn('id', $gameIds)->get()->pluck('id')
+                );
+            }
+
+            $playerGames = $query->get();
+
+            $this->info('Updating ' . $playerGames->count() . ' ' . __res('game', $playerGames->count()) . ' for user [' . $user->id . ':' . $user->username . ']');
+
+            $progressBar = $this->output->createProgressBar($playerGames->count());
+            $progressBar->start();
+
+            // if running in sync mode, just call updatePlayerMetrics once manually after updating
+            // all of the player_games instead of letting it cascade for every player_game updated.
+            $isSync = (config('queue.default') === 'sync');
+
+            foreach ($playerGames as $playerGame) {
+                $this->updatePlayerGameMetrics->execute($playerGame, silent: $isSync);
+                $progressBar->advance();
+            }
+
+            if ($isSync) {
+                $this->updatePlayerMetrics->execute($user);
+            }
+        } else {
+            // No userId was passed. Process all users and games in player_games.
+            $query = PlayerGame::query();
+
+            $rowCount = $query->count();
+            $this->info('Preparing batch jobs to update player game metrics for ' . $rowCount . ' entries.');
+
+            $progressBar = $this->output->createProgressBar($rowCount);
+            $progressBar->start();
+
+            $query->chunkById(100, function ($rows) use ($progressBar) {
+                $jobs = $rows->map(fn ($row) => (new UpdatePlayerGameMetricsJob($row->user_id, $row->game_id))->onQueue('player-game-metrics')
+                )->all();
+
+                Bus::batch($jobs)->onQueue('player-game-metrics')->dispatch();
 
-        if ($isSync) {
-            $this->updatePlayerMetrics->execute($user);
+                $progressBar->advance(count($rows));
+            });
```

Comment:
This took about an hour on my local machine, and it's probably not something we want to do beyond the initial seeding.

We could just seed with a query (eliminating this code), but even that took 7+ minutes, and I don't know if a single 7 minute query is better for the server than several million individual queries.
```
> UPDATE player_games SET achievements_unlocked_softcore = achievements_unlocked - achievements_unlocked_hardcore;
Query OK, 10914277 rows affected (7 min 6.703 sec)
Rows matched: 11641973  Changed: 10914277  Warnings: 0
```

Reply:
That's a good point. If we don't imagine having to run this against the entire table again, then removing this code and doing the update with straight SQL might be best. However, running a single update against such a large table could lock it up during the duration of the update's execution. 

To mitigate this, we may be able to batch the update. Let me do some research, and I can get back to you.

Reply:
Ok, I tested a few things and if we wanted to use SQL for this update we could utilize a temporary stored procedure that batches rows in chunks and commits them in transactions. It's still a lot of data to process, but at least it wouldn't lock quite as much of the table all at once. Here's a sample of what I ran on my local:

```
DROP PROCEDURE IF EXISTS UpdateSoftcoreUnlocks;
SELECT COUNT(*) AS "Rows to Process" FROM player_games WHERE achievements_unlocked_softcore IS NULL;

DELIMITER //

CREATE PROCEDURE UpdateSoftcoreUnlocks()
BEGIN
    DECLARE done INT DEFAULT FALSE;
    DECLARE batch_size INT DEFAULT 10000;
    DECLARE last_id INT DEFAULT 0;
    DECLARE total_rows INT;

    DECLARE cur CURSOR FOR
        SELECT id
        FROM player_games
        WHERE achievements_unlocked_softcore IS NULL
        ORDER BY id
        LIMIT batch_size;

    DECLARE CONTINUE HANDLER FOR NOT FOUND SET done = TRUE;

    SET total_rows = (SELECT COUNT(*) FROM player_games WHERE achievements_unlocked_softcore IS NULL);

    WHILE total_rows > 0 DO
        START TRANSACTION;

        OPEN cur;

        read_loop: LOOP
            FETCH cur INTO last_id;

            IF done THEN
                LEAVE read_loop;
            END IF;

            UPDATE player_games
            SET achievements_unlocked_softcore = achievements_unlocked - achievements_unlocked_hardcore
            WHERE id = last_id;

        END LOOP;

        CLOSE cur;

        COMMIT;

        SET total_rows = (SELECT COUNT(*) FROM player_games WHERE achievements_unlocked_softcore IS NULL AND id > last_id);
        SET done = FALSE;
    END WHILE;
END //

DELIMITER ;

CALL UpdateSoftcoreUnlocks();
DROP PROCEDURE UpdateSoftcoreUnlocks;

SELECT COUNT(*) AS "Rows Processed" FROM player_games WHERE achievements_unlocked_softcore IS NOT NULL;
```
Just let me know which direction you'd like to go in. We could run something like this, just execute a straight SQL update, or utilize artisan and queue a ton of jobs to do individual updates.

Reply:
In the past, I've used bash to iterate. From my notes:
```
for i in $(seq 1 758500); do php artisan ra:platform:player:update-metrics $i; done
```

We could probably do something similar here too.

However this pans out, I want to merge this and run the queries in a separate release from #2466 so the data is ready before the code tries to use it.

Reply:
That sounds good to me. If this PR goes through I'll be sure to update #2466 so it utilizes the new softcore column once everything is ready.

I do like the idea of running a bash loop for the command. Especially since it uses the existing code. As soon as I get a chance, I'll pull the changes to this file.

Reply:
@Jamiras the UpdatePlayerGameMetrics command has been reverted to its original state. I think using bash or SQL is the best option, since this is a one time update. Just let me know if you'd like to discuss how we're going to update the softcore column in `player_games` any further, but resolving the conversation for now.

Reply:
The bash command was processing about 100 users per minute on my VM, meaning it would take ~5 days to fully populate the data. Prod should be faster, but I still expect it to take several hours.

The stored procedure was processing approximately 1000 player_games per second, which would still take about 3 hours.

The best overall approach seems to be to take the 7 minute query listed above and run it in chunks:
```
UPDATE player_games SET achievements_unlocked_softcore=achievements_unlocked-achievements_unlocked_hardcore WHERE id IN (SELECT id FROM player_games
WHERE achievements_unlocked_softcore IS NULL AND id <= 200000);

UPDATE player_games SET achievements_unlocked_softcore=achievements_unlocked-achievements_unlocked_hardcore WHERE id IN (SELECT id FROM player_games
WHERE achievements_unlocked_softcore IS NULL AND id > 200000 AND id <= 400000);
...
```

Each query took about 5 seconds, and we can let the server breathe between them. With ~12m player_games entries, it will take 60 individual queries to update all the records.

Reply:
That sounds like a great idea to me. Just let me know if you need me to generate the script.